Native mode failure: ClassNotFoundException for dataset class generated by table component

I'm using quarkus-jasperreports version 1.0.5 . In JVM dev mode, everything works perfectly, including reports with complex tables and sub-datasets. However, in native mode, a specific .jasper report fails with the following error:

`Caused by: java.lang.ClassNotFoundException: Invoice_Table_Based_tableDataset_f9ab29cba01772b64f47569afcbf590ce32e4ab0fc5835a0bfe9a6086603b6f5
	at net.sf.jasperreports.engine.util.JRClassLoader.loadClass(JRClassLoader.java:29)
	...
	at com.myapp.service.DocumentService.createPdf(DocumentService.java:175)
...
`

This occurs when using a .jrxml report that includes a dataset tag

**Minimal Example Attached:**
I’ve attached the Invoice_Table_Based.jrxml file that reproduces the issue.
Try running it without passing any data — in JVM mode it works, but native mode fails with the above error.

https://github.com/BourassIS/templateExemple/blob/main/Invoice_Table_Based.jrxml

Thanks.

--- a/deployment/src/main/java/io/quarkiverse/jasperreports/deployment/JasperReportsProcessor.java
+++ b/deployment/src/main/java/io/quarkiverse/jasperreports/deployment/JasperReportsProcessor.java
@@ -73,6 +73,7 @@ import io.quarkus.deployment.pkg.builditem.UberJarMergedResourceBuildItem;
 import io.quarkus.jackson.deployment.IgnoreJsonDeserializeClassBuildItem;
 import io.quarkus.logging.Log;
 import net.sf.jasperreports.compilers.ReportExpressionEvaluationData;
+import net.sf.jasperreports.engine.JRDataset;
 import net.sf.jasperreports.engine.JRException;
 import net.sf.jasperreports.engine.JasperCompileManager;
 import net.sf.jasperreports.engine.JasperReport;
@@ -373,19 +374,41 @@ class JasperReportsProcessor extends AbstractJandexProcessor {
             try {
                 String jasperFilePath = reportFile.getPath().toFile().getAbsolutePath();
                 JasperReport report = (JasperReport) JRLoader.loadObject(JRLoader.getLocationInputStream(jasperFilePath));
-                JRReportCompileData reportData = (JRReportCompileData) report.getCompileData();
-                ReportExpressionEvaluationData mainData = (ReportExpressionEvaluationData) reportData
-                        .getMainDatasetCompileData();
-                String reportDataClass = mainData.getCompileName();
-                if (StringUtils.isNotBlank(reportDataClass)) {
-                    byte[] bytes = (byte[]) mainData.getCompileData();
-                    Log.debugf("JasperReport Data Class: %s Size: %d", reportDataClass, bytes.length);
+                JRReportCompileData compileData = (JRReportCompileData) report.getCompileData();
 
-                    reflectiveClassProducer
-                            .produce(ReflectiveClassBuildItem.builder(reportDataClass).constructors().methods()
-                                    .serialization()
-                                    .build());
-                    additionalClasses.produce(new GeneratedClassBuildItem(true, reportDataClass, bytes));
+                // collect all compiled datasets in this report
+                List<ReportExpressionEvaluationData> datasetClasses = new ArrayList<>();
+
+                // add the main data set (should always found)
+                ReportExpressionEvaluationData mainData = (ReportExpressionEvaluationData) compileData
+                        .getMainDatasetCompileData();
+                if (mainData != null) {
+                    datasetClasses.add((ReportExpressionEvaluationData) compileData.getMainDatasetCompileData());
+                }
+
+                // add any sub datasets if found
+                final JRDataset[] datasets = report.getDatasets();
+                if (datasets != null) {
+                    for (JRDataset dataset : datasets) {
+                        ReportExpressionEvaluationData reportData = (ReportExpressionEvaluationData) compileData
+                                .getDatasetCompileData(dataset);
+                        datasetClasses.add(reportData);
+                    }
+                }
+
+                // produce all dataset classes so they are found in native mode
+                for (ReportExpressionEvaluationData reportData : datasetClasses) {
+                    final String reportDataClass = reportData.getCompileName();
+                    if (StringUtils.isNotBlank(reportDataClass)) {
+                        final byte[] bytes = (byte[]) reportData.getCompileData();
+                        Log.debugf("JasperReport Data Class: %s Size: %d", reportDataClass, bytes.length);
+
+                        reflectiveClassProducer
+                                .produce(ReflectiveClassBuildItem.builder(reportDataClass).constructors().methods()
+                                        .serialization()
+                                        .build());
+                        additionalClasses.produce(new GeneratedClassBuildItem(true, reportDataClass, bytes));
+                    }
                 }
             } catch (JRException e) {
                 Log.error("Error loading JasperReport file class.", e);
@@ -463,7 +486,7 @@ class JasperReportsProcessor extends AbstractJandexProcessor {
      *
      * @param config The ReportConfig containing build configuration settings.
      * @return A ReportRootBuildItem representing the default report root directory.
-     *         If build is enabled and a source is specified in the config, it returns that source.
+     *         If a build is enabled and a source is specified in the config, it returns that source.
      *         Otherwise, it returns the default source path defined in ReportConfig.
      */
     @BuildStep
@@ -737,7 +760,7 @@ class JasperReportsProcessor extends AbstractJandexProcessor {
             FileTime inputFileLastModifiedTime = Files.getLastModifiedTime(inputFilePath);
             FileTime outputFileLastModifiedTime = Files.getLastModifiedTime(outputFilePath);
 
-            // Return false if output file is newer or the same as the input file
+            // Return false if the output file is newer or the same as the input file
             if (outputFileLastModifiedTime.compareTo(inputFileLastModifiedTime) >= 0) {
                 Log.debugf("Skipping update, output file %s is up to date.", outputFilePath.toString());
                 return false;
@@ -748,7 +771,8 @@ class JasperReportsProcessor extends AbstractJandexProcessor {
     }
 
     /**
-     * Finds the project root directory. It starts with build target and searches for a directory under that and if that is not
+     * Finds the project root directory. It starts with a build target and searches for a directory under that, and if that is
+     * not
      * found it uses the start directory.
      *
      * @param outputDirectory The output directory path.

--- a/integration-tests/src/main/jasperreports/Invoice_Table_Based.jrxml
+++ b/integration-tests/src/main/jasperreports/Invoice_Table_Based.jrxml
@@ -1,0 +1,163 @@
+<!-- Created with Jaspersoft Studio version 7.0.2.final using JasperReports Library version 7.0.2-31532122640f516c6d17238ae0e41a08113d5ac0  -->
+<jasperReport name="Invoice_Table_Based" language="java" pageWidth="595" pageHeight="842" whenNoDataType="AllSectionsNoDetail" columnWidth="535" leftMargin="20" rightMargin="20" topMargin="20" bottomMargin="20" uuid="4eedbb89-b4f6-4469-9ab6-f642a1688cf7">
+	<property name="template.engine" value="tabular_template"/>
+	<style name="Title" forecolor="#FFFFFF" fontName="Times New Roman" fontSize="50.0" bold="false" pdfFontName="Times-Bold"/>
+	<style name="SubTitle" forecolor="#CCCCCC" fontName="Times New Roman" fontSize="18.0" bold="false" pdfFontName="Times-Roman"/>
+	<style name="Column header" forecolor="#666666" fontName="Times New Roman" fontSize="14.0" bold="true" pdfFontName="Times-Roman"/>
+	<style name="Detail" mode="Transparent" fontName="Times New Roman" pdfFontName="Times-Roman"/>
+	<style name="Row" mode="Transparent" fontName="Times New Roman" pdfFontName="Times-Roman">
+		<conditionalStyle mode="Opaque" backcolor="#EEEFF0">
+			<conditionExpression><![CDATA[$V{REPORT_COUNT}%2 == 0]]></conditionExpression>
+		</conditionalStyle>
+	</style>
+	<style name="Table">
+		<box>
+			<pen lineWidth="1.0" lineColor="#000000"/>
+			<topPen lineWidth="1.0" lineColor="#000000"/>
+			<leftPen lineWidth="1.0" lineColor="#000000"/>
+			<bottomPen lineWidth="1.0" lineColor="#000000"/>
+			<rightPen lineWidth="1.0" lineColor="#000000"/>
+		</box>
+	</style>
+	<style name="Table_TH" mode="Opaque" backcolor="#FFFFFF">
+		<box>
+			<pen lineWidth="0.5" lineColor="#000000"/>
+			<topPen lineWidth="0.5" lineColor="#000000"/>
+			<leftPen lineWidth="0.5" lineColor="#000000"/>
+			<bottomPen lineWidth="0.5" lineColor="#000000"/>
+			<rightPen lineWidth="0.5" lineColor="#000000"/>
+		</box>
+	</style>
+	<style name="Table_CH" mode="Opaque" backcolor="#CACED0">
+		<box>
+			<pen lineWidth="0.5" lineColor="#000000"/>
+			<topPen lineWidth="0.5" lineColor="#000000"/>
+			<leftPen lineWidth="0.5" lineColor="#000000"/>
+			<bottomPen lineWidth="0.5" lineColor="#000000"/>
+			<rightPen lineWidth="0.5" lineColor="#000000"/>
+		</box>
+	</style>
+	<style name="Table_TD" mode="Opaque" backcolor="#FFFFFF">
+		<box>
+			<pen lineWidth="0.5" lineColor="#000000"/>
+			<topPen lineWidth="0.5" lineColor="#000000"/>
+			<leftPen lineWidth="0.5" lineColor="#000000"/>
+			<bottomPen lineWidth="0.5" lineColor="#000000"/>
+			<rightPen lineWidth="0.5" lineColor="#000000"/>
+		</box>
+		<conditionalStyle backcolor="#D8D8D8">
+			<conditionExpression><![CDATA[$V{REPORT_COUNT}%2 == 0]]></conditionExpression>
+		</conditionalStyle>
+	</style>
+	<dataset name="tableDataset" uuid="e467fd5b-ecba-4660-98a4-2d22349de548">
+		<query language="json"><![CDATA[]]></query>
+	</dataset>
+	<query language="json"><![CDATA[]]></query>
+	<title height="196" splitType="Stretch">
+		<element kind="staticText" uuid="57aed2b1-4f4e-40a3-a6ad-54dae8dd4c5a" x="0" y="90" width="84" height="25" fontSize="16.0" bold="true">
+			<text><![CDATA[Bill to:]]></text>
+			<property name="local_mesure_unitheight" value="pixel"/>
+			<property name="com.jaspersoft.studio.unit.height" value="px"/>
+		</element>
+		<element kind="staticText" uuid="33274f5d-5b7e-4be2-a2ff-65d452a618ed" x="0" y="115" width="120" height="15">
+			<text><![CDATA[ABC Company]]></text>
+		</element>
+		<element kind="line" uuid="806ce5df-1219-4876-ae0c-ca7405b1f246" x="0" y="80" width="556" height="1">
+			<property name="local_mesure_unitheight" value="pixel"/>
+			<property name="com.jaspersoft.studio.unit.height" value="px"/>
+		</element>
+		<element kind="staticText" uuid="1d3045ed-03d6-4253-a959-aeeb491b73ca" x="0" y="130" width="120" height="15">
+			<text><![CDATA[Company Address]]></text>
+		</element>
+		<element kind="staticText" uuid="235bf8b6-1e1e-49c7-b59f-7a46c30b8209" x="0" y="145" width="120" height="15">
+			<text><![CDATA[Company state1]]></text>
+		</element>
+		<element kind="staticText" uuid="665f25c2-8663-4951-b699-1ab863bdeb76" x="0" y="160" width="120" height="15">
+			<text><![CDATA[Company state 2]]></text>
+		</element>
+		<element kind="staticText" uuid="b62092ec-9053-4f27-b387-9b25d2909a68" x="410" y="145" width="120" height="15">
+			<text><![CDATA[Ship state1]]></text>
+		</element>
+		<element kind="staticText" uuid="edbd74ac-0cbd-44c9-8a8e-ccf745191172" x="410" y="115" width="120" height="15">
+			<text><![CDATA[Ship name]]></text>
+		</element>
+		<element kind="staticText" uuid="139ebe8d-25b0-411e-986c-270015400874" x="410" y="90" width="84" height="25" fontSize="16.0" bold="true">
+			<text><![CDATA[Ship to:]]></text>
+			<property name="local_mesure_unitheight" value="pixel"/>
+			<property name="com.jaspersoft.studio.unit.height" value="px"/>
+		</element>
+		<element kind="staticText" uuid="dd52166a-1135-4931-adef-51e7972118bc" x="410" y="160" width="120" height="15">
+			<text><![CDATA[Ship state 2]]></text>
+		</element>
+		<element kind="staticText" uuid="2d2f365d-7b12-4a8f-990d-ff528a2660bc" x="410" y="130" width="120" height="15">
+			<text><![CDATA[Ship Address]]></text>
+		</element>
+		<element kind="staticText" uuid="0f86baff-6386-4f3f-b3fe-2388707babe8" x="415" y="10" width="50" height="20" hTextAlign="Right">
+			<text><![CDATA[Date:]]></text>
+			<box rightPadding="4"/>
+		</element>
+		<element kind="textField" uuid="bb10dbe1-0a4f-4722-9953-c163b63cf979" x="465" y="10" width="84" height="20" pattern="EEEEE dd MMMMM yyyy">
+			<expression><![CDATA[new java.util.Date()]]></expression>
+		</element>
+		<element kind="textField" uuid="3836ce65-eca3-4cad-a6de-b1818def0a2b" x="465" y="30" width="84" height="20" pattern="EEEEE dd MMMMM yyyy">
+			<expression><![CDATA["INV0001"]]></expression>
+		</element>
+		<element kind="staticText" uuid="0b3f9342-da78-4cfa-9fc5-2301c4749678" x="415" y="30" width="50" height="20" hTextAlign="Right">
+			<text><![CDATA[Invoice #]]></text>
+			<box rightPadding="4"/>
+		</element>
+		<element kind="staticText" uuid="e622555d-198b-4ccd-a4a1-c59c53304058" x="95" y="0" width="210" height="25" fontSize="19.0" bold="true">
+			<text><![CDATA[Your company name]]></text>
+			<property name="local_mesure_unitheight" value="pixel"/>
+			<property name="com.jaspersoft.studio.unit.height" value="px"/>
+		</element>
+		<element kind="staticText" uuid="b8cc42fd-b569-497d-834d-7e854040ad18" x="95" y="55" width="205" height="15">
+			<text><![CDATA[Phone number, web address, ecc.]]></text>
+		</element>
+		<element kind="staticText" uuid="c226e44d-3bb5-4145-b0b3-903bf1d79fde" x="95" y="25" width="205" height="15">
+			<text><![CDATA[Street address]]></text>
+		</element>
+		<element kind="staticText" uuid="b96f7183-44f8-43ad-ba60-43d8fbf50b6e" x="95" y="40" width="205" height="15">
+			<text><![CDATA[City, street, ZIP code]]></text>
+		</element>
+	</title>
+	<summary height="293" splitType="Stretch">
+		<element kind="staticText" uuid="1a0d7088-5af7-4865-8be1-41ec5f51fb36" positionType="Float" x="0" y="150" width="261" height="101">
+			<text><![CDATA[Your notes here]]></text>
+			<box>
+				<topPen lineWidth="1.1" lineStyle="Dashed"/>
+				<leftPen lineWidth="1.1" lineStyle="Dashed"/>
+				<bottomPen lineWidth="1.1" lineStyle="Dashed"/>
+				<rightPen lineWidth="1.1" lineStyle="Dashed"/>
+			</box>
+		</element>
+		<element kind="staticText" uuid="36aa233d-4305-48e6-974a-1bbf89bb3c8f" positionType="Float" mode="Opaque" x="0" y="263" width="556" height="30" backcolor="#E6E8E9" fontName="Serif" fontSize="9.0" italic="true" hTextAlign="Center" vTextAlign="Middle">
+			<text><![CDATA[THANKYOU FOR YOUR BUSINESS]]></text>
+		</element>
+		<element kind="component" uuid="6bc07188-e6bb-4999-af31-3bb41d702c27" x="0" y="0" width="555" height="130">
+			<component kind="table">
+				<datasetRun uuid="9b5927f7-d57c-409c-bb9a-8a09fa7ea9c4" subDataset="tableDataset">
+					<connectionExpression><![CDATA[$P{REPORT_CONNECTION}]]></connectionExpression>
+				</datasetRun>
+				<column kind="single" uuid="26893cc2-5b56-4a07-a272-910a4dd69e8e" width="160">
+					<columnHeader height="30" style="Table_CH">
+						<element kind="staticText" uuid="6a3ca533-1f68-4ed7-b976-25dbc84d51cf" key="" x="0" y="0" width="160" height="30" forecolor="#000000" fontName="DejaVu Sans" fontSize="12.0" bold="true" style="Column header">
+							<text><![CDATA[]]></text>
+							<box leftPadding="2" style="Column header"/>
+						</element>
+					</columnHeader>
+					<detailCell height="30" style="Table_TD">
+						<element kind="textField" uuid="34f1f311-71f3-43bd-b678-4895aee4b7c2" x="0" y="0" width="160" height="30" fontName="DejaVu Sans" fontSize="12.0" style="Detail">
+							<expression><![CDATA[""]]></expression>
+							<box leftPadding="2" style="Detail"/>
+						</element>
+					</detailCell>
+				</column>
+			</component>
+			<property name="com.jaspersoft.studio.table.style.table_header" value="Table_TH"/>
+			<property name="com.jaspersoft.studio.table.style.column_header" value="Table_CH"/>
+			<property name="com.jaspersoft.studio.table.style.detail" value="Table_TD"/>
+		</element>
+		<property name="local_mesure_unitheight" value="pixel"/>
+	</summary>
+</jasperReport>

--- a/integration-tests/src/main/java/io/quarkiverse/jasperreports/it/JasperReports203Resource.java
+++ b/integration-tests/src/main/java/io/quarkiverse/jasperreports/it/JasperReports203Resource.java
@@ -1,0 +1,72 @@
+/*
+* Licensed to the Apache Software Foundation (ASF) under one or more
+* contributor license agreements.  See the NOTICE file distributed with
+* this work for additional information regarding copyright ownership.
+* The ASF licenses this file to You under the Apache License, Version 2.0
+* (the "License"); you may not use this file except in compliance with
+* the License.  You may obtain a copy of the License at
+*
+*      http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+package io.quarkiverse.jasperreports.it;
+
+import java.io.*;
+import java.util.*;
+
+import jakarta.enterprise.context.ApplicationScoped;
+import jakarta.inject.Inject;
+import jakarta.ws.rs.GET;
+import jakarta.ws.rs.Path;
+import jakarta.ws.rs.ServerErrorException;
+import jakarta.ws.rs.core.HttpHeaders;
+import jakarta.ws.rs.core.MediaType;
+import jakarta.ws.rs.core.Response;
+
+import org.eclipse.microprofile.openapi.annotations.enums.SchemaType;
+import org.eclipse.microprofile.openapi.annotations.media.Content;
+import org.eclipse.microprofile.openapi.annotations.media.Schema;
+import org.eclipse.microprofile.openapi.annotations.responses.APIResponse;
+
+import io.quarkiverse.jasperreports.repository.ReadOnlyStreamingService;
+import io.quarkus.logging.Log;
+import net.sf.jasperreports.engine.*;
+import net.sf.jasperreports.engine.util.JRLoader;
+import net.sf.jasperreports.json.data.JsonDataSource;
+
+@Path("/jasper/203")
+@ApplicationScoped
+public class JasperReports203Resource extends AbstractJasperResource {
+
+    private static final String TEST_REPORT_NAME = "Invoice_Table_Based.jasper";
+
+    @Inject
+    ReadOnlyStreamingService repo;
+
+    @GET
+    @APIResponse(responseCode = "200", description = "PDF generated", content = @Content(mediaType = MediaType.APPLICATION_OCTET_STREAM, schema = @Schema(type = SchemaType.STRING, format = "binary")))
+    public Response generateFromJson() {
+        try {
+            JsonDataSource jsonDataSource = new JsonDataSource(JRLoader.getLocationInputStream("data/sample.json"));
+
+            JasperPrint jasperPrint = JasperFillManager.getInstance(repo.getContext()).fillFromRepo(TEST_REPORT_NAME,
+                    new HashMap<>(), jsonDataSource);
+            ByteArrayOutputStream outputStream = exportPdf(jasperPrint);
+
+            return Response.ok(outputStream.toByteArray())
+                    .header(HttpHeaders.CONTENT_DISPOSITION, "attachment;filename=report.pdf")
+                    .header(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_OCTET_STREAM)
+                    .build();
+
+        } catch (Exception e) {
+            Log.error("Erreur lors de la génération du PDF", e);
+            throw new ServerErrorException(Response.Status.INTERNAL_SERVER_ERROR, e);
+        }
+    }
+
+}

--- a/integration-tests/src/main/resources/data/sample.json
+++ b/integration-tests/src/main/resources/data/sample.json
@@ -1,0 +1,22 @@
+{
+  "tableDataset": [
+    {
+      "item": "Item 001",
+      "description": "Description of item 001",
+      "quantity": 2,
+      "price": 50.0
+    },
+    {
+      "item": "Item 002",
+      "description": "Description of item 002",
+      "quantity": 1,
+      "price": 75.5
+    },
+    {
+      "item": "Item 003",
+      "description": "Description of item 003",
+      "quantity": 4,
+      "price": 20.25
+    }
+  ]
+}

--- a/integration-tests/src/test/java/io/quarkiverse/jasperreports/it/JasperReports203ResourceIT.java
+++ b/integration-tests/src/test/java/io/quarkiverse/jasperreports/it/JasperReports203ResourceIT.java
@@ -1,0 +1,8 @@
+package io.quarkiverse.jasperreports.it;
+
+import io.quarkus.test.junit.QuarkusIntegrationTest;
+
+@QuarkusIntegrationTest
+class JasperReports203ResourceIT extends JasperReportsCsvResourceTest {
+    // Execute the same tests but in packaged mode.
+}

--- a/integration-tests/src/test/java/io/quarkiverse/jasperreports/it/JasperReports203ResourceTest.java
+++ b/integration-tests/src/test/java/io/quarkiverse/jasperreports/it/JasperReports203ResourceTest.java
@@ -1,0 +1,20 @@
+package io.quarkiverse.jasperreports.it;
+
+import static io.restassured.RestAssured.given;
+
+import org.junit.jupiter.api.Test;
+
+import io.quarkus.test.junit.QuarkusTest;
+
+@QuarkusTest
+class JasperReports203ResourceTest {
+
+    @Test
+    void testCsvDatasource() {
+        given()
+                .when().get("/jasper/203")
+                .then()
+                .statusCode(200);
+    }
+
+}

